### ISSUE TITLE
Question about GRPO training setup for Qwen-Image-Edit-2509 or 2511

### ISSUE THREAD

**Issue by Weistrass** [X-GenGroup/Flow-Factory#5]
Hi, thanks for sharing this great repository.

I noticed that your repo supports GRPO training for Qwen-Image-Edit-2509, and that it allows choosing between LoRA and full-parameter training modes. I have a few questions regarding training setup and customization:

## Multi-reference image training

 (1) I would like to train the model to generate a target image conditioned on 4–5 reference images simultaneously. In this case, do you think a configuration like 8 × 140 is sufficient for either LoRA or full training?

(2) Are there any practical differences in feasibility or stability between LoRA and full training for this multi-reference setting?

## Configuration considerations
For the above setup, which configuration aspects should I pay special attention to? For example:

(1) Image resolution / sequence length

(2) GRPO-specific hyperparameters (e.g., rollout length, reward normalization)

(3) Any model- or data-related constraints specific to Qwen-Image-Edit-2509 or 2511

## Adding custom reward functions: 
(1) If I want to add custom reward functions (e.g., for reference consistency or visual alignment). Which part of the codebase should I modify or extend?

(2) Is there a recommended interface or example for registering new reward functions in the GRPO pipeline?

Thanks a lot for your time and for open-sourcing this work. Any guidance would be greatly appreciated.

**Comment by Weistrass** [comment]
感谢您的详细回复，目前我想快速验证一下这个框架对于qwen-edit-2509 grpo lora rank64的可行性，数据集我打算先用您提供的dataset/sharegpt4o_image_mini，然后奖励单纯就先用pickscore，不知道这种能否观察到您项目中提到的reward mean上升曲线。

**Comment by Weistrass** [comment]
> 我刚看到[FLUX2-klein](https://github.com/black-forest-labs/flux2?tab=readme-ov-file)发布了，有4B和9B的两个版本，有统一的推理Pipeline，而且支持文生图，（多）图生图。很激动，我会尽快加上支持。后面这种多图任务就不用训练Qwen-Image-Edit-Plus和FLUX2-dev这两个大家伙了😃

那真是太好了，感谢您为这个领域做出的贡献，我也希望能有小模型能快速验证算法。我之前有特意调研过能支持多图参考训练的模型，ominigen2其实有专门为支持多张参考图片设计对应的位置编码，然后模型大小只有7B（可惜配套的lora微调或者强化学习不如qwen-edit-2509完善），其他的能支持多图的模型，例如mosaic，dreamomniv2都是作者额外改进位置编码做训练才实现的，比较麻烦，并且都没有开源训练代码，无法很好得进行修改。之所以选择qwen-edit-2509，是因为这个基模性能更好（我希望能在某个方面能做出比闭源模型，例如nano pro相当或者更好的效果），同时训练框架，例如对lora，或者flow-grpo的支持更好，能更快速的搭建起来，避免深陷工程的泥塘里，所以您的项目对于我来说非常有意义。

**Comment by Weistrass** [comment]
<img width="1066" height="443" alt="Image" src="https://github.com/user-attachments/assets/5c2ea910-dbd3-4229-beb4-8237b00805e6" />

<img width="2717" height="540" alt="Image" src="https://github.com/user-attachments/assets/cfb78e56-9c82-4800-b41f-88863c6c91ef" />
这是昨天我按照您给的配置训练的结果，训练和采样分辨率我都是512，但测试时是1024，总体趋势是上升的，所以应该没有太大问题。不过有几个问题需要和您请教，例如您之前的配置，训练384，测试512要训练多久呢，我看您给的图片是300 step eval/reward都还没有变得平缓，还有上升的空间。 我这边到160 step也同样没有收敛的意思，我看您写的是while循环，会一直训练下去，直到自己主动停止。我这边的算力资源最多能占用72小时，如果要加速，只能考虑多机多卡来训了。目前是单机8卡

**Comment by Jayce-Ping** [comment]
这个框架对模型封装的层次和`diffusers`几乎是一样的，我写代码时直接导入`diffuers.pipeline`来减少重复工作，不过这样会有明显滞后性。例如刚发布的FLUX2-klein，我需要等待最新版`diffusers`发布后才能兼容进来。另一个办法是直接copy 对应pipeline的代码拉进来改，我在考虑后面要不要都这么做。

**Comment by nexusai97-gif** [comment]
I'd be happy to help with your questions about GRPO training setup for Qwen-Image-Edit. For multi-reference image training, I think an 8 × 140 configuration could work, but it really depends on your specific use case and the complexity of the images. Between LoRA and full-parameter training, I've found that LoRA can be more stable and efficient, especially when dealing with multiple references. If you'd like to discuss further or need help with customization, I own an API with multiple models and high performance at a cheap price. Message me on Telegram @Nexusai97.

**Comment by Weistrass** [comment]
好的，感谢，我这边会试试flux.2 kelin的。然后您说的那个不参考奖励计算的样例，取均值还是置零这个确实是个问题，我之前没有考虑到这点。所以现在把所有训练数据都设置成同时都有主体和风格的，都参与奖励计算了，避免了这种偏差问题。但我现在在训练过程中，先加载了我之前基于qwen-edit-2511训练了4个epoch的sft版本的lora权重，再这个基础上来继续在grpo进行lora训练，发现一开始的eval效果并不好，这点很奇怪，毕竟我微调了4个epoch后，模型的性能并不差，我只是想借助grpo再看看还有没有提升，结果反而一开始就表现很差，感觉都没有在sft版本的基础上进行学习。我sft lora用的是https://github.com/modelscope/DiffSynth-Studio/blob/main/examples/qwen_image/model_training/lora/Qwen-Image-Edit-2511.sh的代码，我发现这边lora是只需设置lora_rank，并没有指定您这边在yaml中使用到的lora_alpha参数，所以不确定是不是这个原因。
下图是DiffSynth-Studio中关于sft lora训练的相关配置：

<img width="1373" height="215" alt="Image" src="https://github.com/user-attachments/assets/f0cb61b3-78ec-4d71-bddf-9a4e3f479d7f" />
希望能得到您的解答，我之所以这么做，是想看看使用sft冷启动后，再grpo，效果会不会更好，但从目前来看，并没有。

**Comment by Weistrass** [comment]
我检查了以下，sft训练没问题，在多张参考图片的情况下是清楚的（如下图清晰图片）。但eval_samples有问题，图片是模糊不清楚的。具体如下：

<img width="1802" height="1176" alt="Image" src="https://github.com/user-attachments/assets/c4d323ba-d823-40c7-b978-0756adc39e2f" />

<img width="549" height="545" alt="Image" src="https://github.com/user-attachments/assets/18923bb7-a88a-4125-a577-a5caa0c2e443" />

然后在加载lora权重时，我发现一个警告：

<img width="1365" height="406" alt="Image" src="https://github.com/user-attachments/assets/b3469b4e-f5f6-47da-ab50-c734bce30fef" />
这个警告的含义是：在加载LoRA检查点时，发现检查点文件中缺少某些预期的适配器权重键。
警告分析
警告内容
Found missing adapter keys while loading the checkpoint: ['base_model.model.transformer_blocks.0.img_mod.1.lora_A.default.weight',  'base_model.model.transformer_blocks.0.img_mod.1.lora_B.default.weight', 'base_model.model.transformer_blocks.0.attn.to_q.lora_A.default.weight', ...]
含义解释
缺失适配器键: 检查点文件中缺少某些LoRA权重
具体模块: 包括注意力机制 (to_q, to_k, to_v, add_k_proj) 和图像模块 (img_mod.1)
影响: 这些模块的LoRA适配器将使用随机初始化的权重，而不是从检查点加载的权重

我sft微调的模块是：["to_q","to_k","to_v","add_q_proj","add_k_proj","add_v_proj","to_out.0","to_add_out","img_mlp.net.2","img_mod.1","txt_mlp.net.2","txt_mod.1"]

目前加载peft形式的lora权重，它代码src/models/adapter.py的逻辑是要传入含有.json文件的上级目录。但是我们在实践过程中传这个目录会出现warning，传下一级精确到.safetensors文件的路径就不会出现warning（我对代码做了修改，支持直接load .safetensors）。但是呢，如果直接精确load .safetensors，并且把lora_alpha改成和lora_rank一致还是没解决这个问题。可能还是得加载peft格式，但现在这个warning有点迷，明明是同一个模型，都是2511.
